File: javascript/oop-patterns/behavioral-patterns/state/state.js
// Контекст - объект, используемый клиентом. Он хранит в себе ссылку на объект
// текущего состояния и меняет своё поведение в зависимости от этого
// состояния
class Context {
  state;

  constructor(state) {
    this.transitionTo(state);
  }

  transitionTo(state) {
    console.log(`Контекст: Переход к ${state.constructor.name}.`);
    this.state = state;
    this.state.setContext(this);
  }

  request1() {
    this.state ? this.state.handle1() : null;
  }

  request2() {
    this.state ? this.state.handle2() : null;
  }
}

// Базовое состояние содержит методы, которые должны присутствовать в каждом
// объекте состояния, а так же ссылку на контекст для того, чтобы
// иметь возможность его изменить
class State {
  context;

  setContext(context) {
    this.context = context;
  }

  handle1() {}

  handle2() {}
}

class ConcreteStateA extends State {
  handle1() {
    console.log('ConcreteStateA обрабатывает request1.');
    console.log('ConcreteStateA хочет изменить состояние контекста.');
    this.context ? this.context.transitionTo(new ConcreteStateB()) : null;
  }

  handle2() {
    console.log('ConcreteStateA обрабатывает request2.');
  }
}

class ConcreteStateB extends State {
  handle1() {
    console.log('ConcreteStateB обрабатывает request1.');
  }

  handle2() {
    console.log('ConcreteStateB обрабатывает request2.');
    console.log('ConcreteStateB хочет изменить состояние контекста.');
    this.context ? this.context.transitionTo(new ConcreteStateA()) : null;
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
const context = new Context(new ConcreteStateA());
context.request1();
context.request2();

File: go/oop-patterns/behavioral-patterns/state/state.go
package main

import "fmt"

// Контекст - объект, используемый клиентом. Он хранит в себе ссылку на объект
// текущего состояния и меняет своё поведение в зависимости от этого
// состояния
type context struct {
	state iState
}

func (c *context) transitionTo(s iState) {
	fmt.Println("Контекст: Переход к", s.getName())
	c.state = s
	s.setContext(*c)
}

func (c *context) request1() {
	c.state.handle1()
}

func (c *context) request2() {
	c.state.handle2()
}

// Базовое состояние содержит методы, которые должны присутствовать в каждом
// объекте состояния, а так же ссылку на контекст для того, чтобы
// иметь возможность его изменить
type state struct {
	context *context
	name    string
}

type iState interface {
	getName() string
	setContext(context context)
	handle1()
	handle2()
}

func (s *state) setContext(context context) {
	s.context = &context
}
func (s *state) handle1() {
}
func (s *state) handle2() {
}
func (s *state) getName() string {
	return s.name
}

type concreteStateA struct {
	state
}

func (s *concreteStateA) handle1() {
	fmt.Println("ConcreteStateA обрабатывает request1.")
	fmt.Println("ConcreteStateA хочет изменить состояние контекста.")
	s.context.transitionTo(&concreteStateB{state{name: "concreteStateB"}})
}

func (s *concreteStateA) handle2() {
	fmt.Println("ConcreteStateA обрабатывает request2.")

}

type concreteStateB struct {
	state
}

func (s *concreteStateB) handle1() {
	fmt.Println("ConcreteStateB обрабатывает request2.")
}

func (s *concreteStateB) handle2() {
	fmt.Println("ConcreteStateB обрабатывает request2.")
	fmt.Println("ConcreteStateB хочет изменить состояние контекста.")
	s.context.transitionTo(&concreteStateA{state{name: "concreteStateA"}})
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func main() {
	context := context{}
	initialState := &concreteStateA{state{name: "concreteStateA", context: &context}}
	context.state = initialState
	context.request1()
	context.request2()
}
